runtime/runtime/src/congestion_control.rs

```diff
@@ -656,29 +676,61 @@ impl DelayedReceiptQueueWrapper {
         Ok(())
     }
 
+    // With ReshardingV3, it's possible for a chunk to have delayed receipts that technically belong
+    // the sibling shard before a resharding event.
+    // Here, we filter all the receipts that don't belong to the current shard_id.
+    //
+    // The function follows the guidelines of standard iterator filter function
+    // We return true if we should retain the receipt and false if we should filter it.
+    fn receipt_filter_fn(&self, receipt: &ReceiptOrStateStoredReceipt) -> bool {
+        if !ProtocolFeature::SimpleNightshadeV4.enabled(self.protocol_version) {
+            return true;
+        }
+        let receiver_id = receipt.get_receipt().receiver_id();
+        let receipt_shard_id = self
+            .epoch_info_provider
+            .account_id_to_shard_id(receiver_id, &self.epoch_id)
+            .expect("account_id_to_shard_id should never fail");
+        receipt_shard_id == self.shard_id
+    }
+
     pub(crate) fn pop(
         &mut self,
         trie_update: &mut TrieUpdate,
         config: &RuntimeConfig,
     ) -> Result<Option<ReceiptOrStateStoredReceipt>, RuntimeError> {
-        let receipt = self.queue.pop_front(trie_update)?;
-        if let Some(receipt) = &receipt {
-            let delayed_gas = receipt_congestion_gas(receipt, &config)?;
-            let delayed_bytes = receipt_size(receipt)? as u64;
+        // While processing receipts, we need to keep track of the gas and bytes
+        // even for receipts that may be filtered out due to a resharding event
+        while let Some(receipt) = self.queue.pop_front(trie_update)? {
+            let delayed_gas = receipt_congestion_gas(&receipt, &config)?;
+            let delayed_bytes = receipt_size(&receipt)? as u64;
             self.removed_delayed_gas = safe_add_gas(self.removed_delayed_gas, delayed_gas)?;
             self.removed_delayed_bytes = safe_add_gas(self.removed_delayed_bytes, delayed_bytes)?;
+
+            // Track gas and bytes for receipt above and return only receipt that belong to the shard.
+            if self.receipt_filter_fn(&receipt) {
+                return Ok(Some(receipt));
+            }
```

Comment:
This loop here may be bypassing some of the limits imposed on receipt processing. The proper solution may need to be done on a higher level but that seems a bit nasty. I'd be willing to risk it but I'm curious about your and @jancionear opinions on that.

Reply:
That's a valid issue. I completely missed it. Let me also think about it a bit

Reply:
btw, may be worth merging as is and just adding a todo to unblock others

Reply:
Added TODO comment